llvm/lib/MC/MCDwarf.cpp

```diff
@@ -185,29 +190,27 @@ void MCDwarfLineTable::emitOne(
   for (const MCDwarfLineEntry &LineEntry : LineEntries) {
     MCSymbol *Label = LineEntry.getLabel();
     const MCAsmInfo *asmInfo = MCOS->getContext().getAsmInfo();
-    if (LineEntry.IsEndEntry) {
+    if (LineEntry.IsEndEntry || LineEntry.IsEndOfFunction) {
       MCOS->emitDwarfAdvanceLineAddr(INT64_MAX, LastLabel, Label,
                                      asmInfo->getCodePointerSize());
+      unsigned PrevFileNum = FileNum;
       init();
+      // File number doesn't need to be reset and extra DW_LNS_set_file
+      // directives reduces deduplication
+      if (LineEntry.IsEndOfFunction)
+        FileNum = PrevFileNum;
       EndEntryEmitted = true;
       continue;
     }
-    if (LineEntry.IsEndOfFunction) {
-      MCOS->emitInt8(char(dwarf::DW_LNS_extended_op));
-      MCOS->emitInt8(char(1));
-      MCOS->emitInt8(char(dwarf::DW_LNE_end_sequence));
-      init();
-      continue;
-    }
 
     int64_t LineDelta = static_cast<int64_t>(LineEntry.getLine()) - LastLine;
 
-    if (FileNum != LineEntry.getFileNum() || LineEntry.IsEndOfFunction) {
+    if (FileNum != LineEntry.getFileNum()) {
```

Comment:
This looks like there could be some NFC refactoring done upstream first? Such as moving the LineEntry.IsEndOfFunction higher up? Might make maintaining the patch easier.